#!/usr/bin/env python3
"""
AIæœåŠ¡é›†æˆæµ‹è¯•è„šæœ¬
"""

import grpc
import sys
import os
import json
import time
import asyncio
from concurrent.futures import ThreadPoolExecutor

# æ·»åŠ å½“å‰ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import intelligence_pb2
import intelligence_pb2_grpc

class AIServiceIntegrationTest:
    def __init__(self):
        self.total_tests = 0
        self.passed_tests = 0
        self.failed_tests = 0
        self.channel = None
        self.stub = None
    
    def test_result(self, condition, message):
        """è®°å½•æµ‹è¯•ç»“æœ"""
        self.total_tests += 1
        if condition:
            print(f"âœ… {message}")
            self.passed_tests += 1
        else:
            print(f"âŒ {message}")
            self.failed_tests += 1
    
    def setup_grpc_connection(self):
        """å»ºç«‹gRPCè¿æ¥"""
        try:
            self.channel = grpc.insecure_channel('localhost:50051')
            grpc.channel_ready_future(self.channel).result(timeout=10)
            self.stub = intelligence_pb2_grpc.IntelligenceServiceStub(self.channel)
            self.test_result(True, "gRPCè¿æ¥å»ºç«‹æˆåŠŸ")
            return True
        except Exception as e:
            self.test_result(False, f"gRPCè¿æ¥å»ºç«‹å¤±è´¥: {e}")
            return False
    
    def test_grpc_service_availability(self):
        """æµ‹è¯•gRPCæœåŠ¡å¯ç”¨æ€§"""
        print("\n1. gRPCæœåŠ¡å¯ç”¨æ€§æµ‹è¯•")
        print("--------------------")
        
        if not self.setup_grpc_connection():
            return
        
        # æµ‹è¯•æœåŠ¡æ˜¯å¦å“åº”
        try:
            # åˆ›å»ºä¸€ä¸ªç®€å•çš„è¯·æ±‚
            request = intelligence_pb2.InterpretRequest()
            request.user_text = "hello"
            request.session_id = "test-session"
            
            # è®¾ç½®è¶…æ—¶æ—¶é—´
            response = self.stub.InterpretText(request, timeout=10)
            self.test_result(True, "gRPCæœåŠ¡å“åº”æµ‹è¯•")
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.UNAVAILABLE:
                self.test_result(False, "gRPCæœåŠ¡ä¸å¯ç”¨")
            else:
                self.test_result(True, f"gRPCæœåŠ¡å“åº” (çŠ¶æ€: {e.code()})")
        except Exception as e:
            self.test_result(False, f"gRPCæœåŠ¡æµ‹è¯•å¼‚å¸¸: {e}")
    
    def test_text_interpretation(self):
        """æµ‹è¯•æ–‡æœ¬è§£é‡ŠåŠŸèƒ½"""
        print("\n2. æ–‡æœ¬è§£é‡ŠåŠŸèƒ½æµ‹è¯•")
        print("------------------")
        
        test_cases = [
            {
                "input": "æ›´æ–°æˆ‘çš„ç”µè¯å·ç ä¸º13800138000",
                "expected_intent": "update_phone_number",
                "description": "ç”µè¯å·ç æ›´æ–°æ„å›¾è¯†åˆ«"
            },
            {
                "input": "è°æ˜¯æˆ‘çš„ç»ç†ï¼Ÿ",
                "expected_intent": "get_employee_manager",
                "description": "æŸ¥è¯¢ç»ç†æ„å›¾è¯†åˆ«"
            },
            {
                "input": "Hello there",
                "expected_intent": "no_intent_detected",
                "description": "æ— æ„å›¾è¯†åˆ«"
            }
        ]
        
        for i, case in enumerate(test_cases, 1):
            try:
                request = intelligence_pb2.InterpretRequest()
                request.user_text = case["input"]
                request.session_id = f"test-session-{i}"
                
                response = self.stub.InterpretText(request, timeout=15)
                
                # éªŒè¯å“åº”æ ¼å¼
                self.test_result(hasattr(response, 'intent'), f"{case['description']} - å“åº”åŒ…å«intentå­—æ®µ")
                self.test_result(hasattr(response, 'structured_data_json'), f"{case['description']} - å“åº”åŒ…å«structured_data_jsonå­—æ®µ")
                
                # éªŒè¯JSONæ ¼å¼
                if response.structured_data_json:
                    try:
                        json.loads(response.structured_data_json)
                        self.test_result(True, f"{case['description']} - JSONæ ¼å¼æ­£ç¡®")
                    except json.JSONDecodeError:
                        self.test_result(False, f"{case['description']} - JSONæ ¼å¼é”™è¯¯")
                
            except grpc.RpcError as e:
                self.test_result(False, f"{case['description']} - gRPCé”™è¯¯: {e.code()}")
            except Exception as e:
                self.test_result(False, f"{case['description']} - å¼‚å¸¸: {e}")
    
    def test_concurrent_requests(self):
        """æµ‹è¯•å¹¶å‘è¯·æ±‚å¤„ç†"""
        print("\n3. å¹¶å‘è¯·æ±‚å¤„ç†æµ‹è¯•")
        print("------------------")
        
        def make_request(session_id):
            try:
                request = intelligence_pb2.InterpretRequest()
                request.user_text = f"æµ‹è¯•å¹¶å‘è¯·æ±‚ {session_id}"
                request.session_id = f"concurrent-test-{session_id}"
                
                response = self.stub.InterpretText(request, timeout=20)
                return True
            except Exception as e:
                print(f"å¹¶å‘è¯·æ±‚ {session_id} å¤±è´¥: {e}")
                return False
        
        # å¹¶å‘æ‰§è¡Œ5ä¸ªè¯·æ±‚
        with ThreadPoolExecutor(max_workers=5) as executor:
            futures = [executor.submit(make_request, i) for i in range(5)]
            results = [future.result() for future in futures]
        
        success_count = sum(results)
        self.test_result(success_count >= 3, f"å¹¶å‘è¯·æ±‚å¤„ç† ({success_count}/5 æˆåŠŸ)")
    
    def test_error_handling(self):
        """æµ‹è¯•é”™è¯¯å¤„ç†"""
        print("\n4. é”™è¯¯å¤„ç†æµ‹è¯•")
        print("--------------")
        
        # æµ‹è¯•ç©ºè¾“å…¥ - åº”è¯¥è¢«æ‹’ç»å¹¶è¿”å›é”™è¯¯
        try:
            request = intelligence_pb2.InterpretRequest()
            request.user_text = ""
            request.session_id = "empty-test"
            
            response = self.stub.InterpretText(request, timeout=10)
            # å¦‚æœæ²¡æœ‰æŠ›å‡ºå¼‚å¸¸ï¼Œè¯´æ˜ç©ºè¾“å…¥è¢«å¤„ç†äº†ï¼Œè¿™ä¸æ˜¯æœŸæœ›çš„è¡Œä¸º
            self.test_result(False, "ç©ºè¾“å…¥åº”è¯¥è¢«æ‹’ç»")
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.INVALID_ARGUMENT:
                self.test_result(True, "ç©ºè¾“å…¥æ­£ç¡®è¢«æ‹’ç»")
            else:
                self.test_result(False, f"ç©ºè¾“å…¥å¤„ç†å¼‚å¸¸: {e.code()}")
        except Exception as e:
            self.test_result(False, f"ç©ºè¾“å…¥å¤„ç†å¤±è´¥: {e}")
        
        # æµ‹è¯•è¶…é•¿è¾“å…¥
        try:
            request = intelligence_pb2.InterpretRequest()
            request.user_text = "x" * 10000  # è¶…é•¿æ–‡æœ¬
            request.session_id = "long-test"
            
            response = self.stub.InterpretText(request, timeout=30)
            # å¦‚æœå¤„ç†æˆåŠŸï¼Œè¯´æ˜ç³»ç»Ÿèƒ½å¤„ç†è¶…é•¿è¾“å…¥
            self.test_result(True, "è¶…é•¿è¾“å…¥å¤„ç†")
        except grpc.RpcError as e:
            if e.code() == grpc.StatusCode.INVALID_ARGUMENT:
                self.test_result(True, "è¶…é•¿è¾“å…¥æ­£ç¡®è¢«æ‹’ç»")
            else:
                self.test_result(False, f"è¶…é•¿è¾“å…¥å¤„ç†å¼‚å¸¸: {e.code()}")
        except Exception as e:
            self.test_result(False, f"è¶…é•¿è¾“å…¥å¤„ç†å¤±è´¥: {e}")
    
    def test_response_time(self):
        """æµ‹è¯•å“åº”æ—¶é—´"""
        print("\n5. å“åº”æ—¶é—´æµ‹è¯•")
        print("--------------")
        
        response_times = []
        
        for i in range(3):
            try:
                start_time = time.time()
                
                request = intelligence_pb2.InterpretRequest()
                request.user_text = "å¿«é€Ÿå“åº”æµ‹è¯•"
                request.session_id = f"speed-test-{i}"
                
                response = self.stub.InterpretText(request, timeout=30)
                
                end_time = time.time()
                response_time = (end_time - start_time) * 1000  # è½¬æ¢ä¸ºæ¯«ç§’
                response_times.append(response_time)
                
            except Exception as e:
                print(f"å“åº”æ—¶é—´æµ‹è¯• {i+1} å¤±è´¥: {e}")
        
        if response_times:
            avg_time = sum(response_times) / len(response_times)
            self.test_result(avg_time < 10000, f"å¹³å‡å“åº”æ—¶é—´ ({avg_time:.0f}ms)")
            self.test_result(max(response_times) < 30000, f"æœ€å¤§å“åº”æ—¶é—´ ({max(response_times):.0f}ms)")
        else:
            self.test_result(False, "å“åº”æ—¶é—´æµ‹è¯•å¤±è´¥")
    
    def test_session_management(self):
        """æµ‹è¯•ä¼šè¯ç®¡ç†"""
        print("\n6. ä¼šè¯ç®¡ç†æµ‹è¯•")
        print("--------------")
        
        session_id = "session-management-test"
        
        # å‘é€ç¬¬ä¸€ä¸ªè¯·æ±‚
        try:
            request1 = intelligence_pb2.InterpretRequest()
            request1.user_text = "æˆ‘æƒ³æ›´æ–°ä¸ªäººä¿¡æ¯"
            request1.session_id = session_id
            
            response1 = self.stub.InterpretText(request1, timeout=15)
            self.test_result(True, "ä¼šè¯ç¬¬ä¸€ä¸ªè¯·æ±‚")
            
            # å‘é€ç›¸å…³çš„ç¬¬äºŒä¸ªè¯·æ±‚
            request2 = intelligence_pb2.InterpretRequest()
            request2.user_text = "æ›´æ–°ç”µè¯å·ç "
            request2.session_id = session_id
            
            response2 = self.stub.InterpretText(request2, timeout=15)
            self.test_result(True, "ä¼šè¯ç›¸å…³è¯·æ±‚")
            
        except Exception as e:
            self.test_result(False, f"ä¼šè¯ç®¡ç†æµ‹è¯•å¤±è´¥: {e}")
    
    def cleanup(self):
        """æ¸…ç†èµ„æº"""
        if self.channel:
            self.channel.close()
    
    def run_all_tests(self):
        """è¿è¡Œæ‰€æœ‰æµ‹è¯•"""
        print("ğŸ° Cube Castle - AIæœåŠ¡é›†æˆæµ‹è¯•")
        print("==============================")
        
        self.test_grpc_service_availability()
        self.test_text_interpretation()
        self.test_concurrent_requests()
        self.test_error_handling()
        self.test_response_time()
        self.test_session_management()
        
        self.cleanup()
        
        print("\n==============================")
        print("AIæœåŠ¡é›†æˆæµ‹è¯•å®Œæˆï¼")
        print(f"æ€»è®¡: {self.total_tests} é¡¹æµ‹è¯•")
        print(f"âœ… é€šè¿‡: {self.passed_tests} é¡¹")
        print(f"âŒ å¤±è´¥: {self.failed_tests} é¡¹")
        success_rate = (self.passed_tests / self.total_tests * 100) if self.total_tests > 0 else 0
        print(f"æˆåŠŸç‡: {success_rate:.1f}%")
        print("==============================")
        
        return self.failed_tests == 0

if __name__ == "__main__":
    test_runner = AIServiceIntegrationTest()
    success = test_runner.run_all_tests()
    sys.exit(0 if success else 1)